options/logger_registry.go

```diff
@@ -0,0 +1,112 @@
+package options
+
+import (
+	"encoding/json"
+	"sync"
+
+	"github.com/mongodb/grip/send"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+// LoggerRegistry is an interface that stores reusable logger factories.
+type LoggerRegistry interface {
+	Register(string, LoggerFactory)
+	Check(string) bool
+	Names() []string
+	Resolve(string) (LoggerFactory, bool)
+}
+
+type basicLoggerRegistry struct {
+	mu        sync.RWMutex
+	factories map[string]LoggerFactory
+}
+
+func NewBasicLoggerRegistry() LoggerRegistry {
+	return &basicLoggerRegistry{
+		factories: map[string]LoggerFactory{},
+	}
+}
+
+func (r *basicLoggerRegistry) Register(name string, factory LoggerFactory) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.factories[name] = factory
+}
+
+func (r *basicLoggerRegistry) Check(name string) bool {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	_, ok := r.factories[name]
+	return ok
+}
+
+func (r *basicLoggerRegistry) Names() []string {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	names := []string{}
+	for name := range r.factories {
+		names = append(names, name)
+	}
+
+	return names
+}
+
+func (r *basicLoggerRegistry) Resolve(name string) (LoggerFactory, bool) {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	factory, ok := r.factories[name]
+	return factory, ok
+}
+
+// LogConfigFormat describes the format of the log configuration.
+type LogConfigFormat string
+
+const (
+	LogConfigFormatBSON LogConfigFormat = "BSON"
+	LogConfigFormatJSON LogConfigFormat = "JSON"
+)
+
+func (f LogConfigFormat) Validate() error {
+	switch f {
+	case LogConfigFormatBSON, LogConfigFormatJSON:
+	default:
+		return errors.New("unknown log config format")
+	}
+
+	return nil
+}
+
+// Unmarshal unmarshals the data into out based on the log config format.
+func (f LogConfigFormat) Unmarshal(data []byte, out interface{}) error {
+	switch f {
+	case LogConfigFormatBSON:
+		if err := bson.Unmarshal(data, out); err != nil {
+			return errors.Wrapf(err, "could not render '%s' input into '%s'", data, out)
+
+		}
+
+		return nil
+	case LogConfigFormatJSON:
+		if err := json.Unmarshal(data, out); err != nil {
+			return errors.Wrapf(err, "could not render '%s' input into '%s'", data, out)
+
+		}
+
+		return nil
+	default:
+		return errors.Errorf("unsupported format '%s'", f)
+	}
+}
+
+// LoggerProducer produces a Logger interface backed by a grip logger.
+type LoggerProducer interface {
+	Configure() (send.Sender, error)
```

Comment:
you added a validate method here, and I like that, I think it should be part of the interface. 

I also think that you have to specify the type here here, so that you can write an implementation in the marshalers that do the right thing. (see my Set() method.)

Reply:
I think the validate method should be build in to the Configure() since it already returns an error (and its up to the implementation to decided what needs to be validated, if anything at all). I changed the LoggerRegistry to use the LoggerProducer.Type() method as the name (so we don't need to check if after the fact).